promql/promqltest/test.go

```diff
@@ -693,29 +792,77 @@ func (ev *evalCmd) expectMetric(pos int, m labels.Labels, vals ...parser.Sequenc
 	ev.expected[h] = entry{pos: pos, vals: vals}
 }
 
+// validateExpectedAnnotations validates expected messages and regex match actual annotations.
+func validateExpectedAnnotations(expr string, expectedAnnotations []expectCmd, actualAnnotations []string, line int, annotationType string) error {
+	if len(expectedAnnotations) == 0 {
+		return nil
+	}
+
+	if len(actualAnnotations) == 0 {
+		return fmt.Errorf("expected %s annotations but none were found for query %q (line %d)", annotationType, expr, line)
+	}
+
+	// Check if all expected annotations are found in actual.
+	for _, e := range expectedAnnotations {
+		matchFound := slices.ContainsFunc(actualAnnotations, e.CheckMatch)
+		if !matchFound {
+			return fmt.Errorf("expected %s annotation %q but no matching annotation was found for query %q (line %d)", annotationType, e.String(), expr, line)
```

Comment:
Might be good to make it clear when we're expecting a pattern or a string, so that the error messages are unambiguous, especially if pattern / string contains special characters for regexes

eg. at the moment, we'll get errors like `expected info annotation "foo" but no matching annotation was found for query "blah" (line 123)`, but something like `expected info annotation matching pattern "foo" but no matching annotation was found for query "blah" (line 123)` would be clearer

(same applies to other error messages below)

Reply:
Yeah, it would be more readable and would be easy to debug.

Reply:
Done. Please check.

Reply:
LGTM, left one suggestion below.